File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int len = nums.size();
        vector<int> res(len, 1);

        for (int i = 1; i < len; i++) {
            res[i] = res[i - 1] * nums[i - 1];
        }

        int postfix = 1;
        for (int i = len - 1; i >= 0; i--) {
            res[i] *= postfix;
            postfix *= nums[i];
        }

        return res;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]: 
        nums.sort()
        left, middle = 0, 1
        result = list()
        maximum_right = len(nums) - 1
        while nums[left] <= 0 and middle < maximum_right:
            right = maximum_right
            while nums[left] + nums[middle] + nums[right] > 0 and right > middle:
                right -= 1
                maximum_right = right
            while middle < right:
                current_sum = nums[left] + nums[middle] + nums[right]
                if current_sum == 0:
                    result.append([nums[left], nums[middle], nums[right]])
                    middle += 1
                    right -= 1
                    while nums[middle] == nums[middle - 1] and middle < right :
                        middle += 1
                elif current_sum > 0:
                    right -= 1
                else:
                    middle += 1
            left += 1
            while nums[left] == nums[left - 1] and left < middle:
                left += 1
            middle = left + 1

        return result